toolchain/check/testdata/where_expr/constraints.carbon

```diff
@@ -19,23 +19,101 @@ interface I {
   let Second:! J;
 }
 
-fn Equal(T:! I where .Member = {});
-
 fn EqualEqual(U:! I where .Self == ());
 
 fn Impls(V:! J where .Self impls I);
 
-fn And(W:! I where .Second impls I and .Member = .Second);
+fn And(W:! I where .Self impls J and .Member == ());
+
+// --- fail_todo_equal_constraint.carbon
+
+library "[[@TEST_NAME]]";
+
+interface N {
+  let P:! type;
+}
+
+// CHECK:STDERR: fail_todo_equal_constraint.carbon:[[@LINE+7]]:27: error: cannot implicitly convert from `{}` to `<associated type in N>`
+// CHECK:STDERR: fn Equal(T:! N where .P = {});
+// CHECK:STDERR:                           ^~
+// CHECK:STDERR: fail_todo_equal_constraint.carbon:[[@LINE+4]]:27: note: type `{}` does not implement interface `ImplicitAs`
+// CHECK:STDERR: fn Equal(T:! N where .P = {});
+// CHECK:STDERR:                           ^~
+// CHECK:STDERR:
+fn Equal(T:! N where .P = {});
+
+// --- fail_todo_associated_type_impls.carbon
 
-// --- todo_check_constraints.carbon
+library "[[@TEST_NAME]]";
+
+interface L {}
+interface M {}
+
+interface K {
+  let Associated:! L;
+}
+
+// CHECK:STDERR: fail_todo_associated_type_impls.carbon:[[@LINE+7]]:36: error: cannot implicitly convert from `<associated L in K>` to `type`
+// CHECK:STDERR: fn AssociatedTypeImpls(W:! K where .Associated impls M);
+// CHECK:STDERR:                                    ^~~~~~~~~~~
+// CHECK:STDERR: fail_todo_associated_type_impls.carbon:[[@LINE+4]]:36: note: type `<associated L in K>` does not implement interface `ImplicitAs`
+// CHECK:STDERR: fn AssociatedTypeImpls(W:! K where .Associated impls M);
+// CHECK:STDERR:                                    ^~~~~~~~~~~
+// CHECK:STDERR:
+fn AssociatedTypeImpls(W:! K where .Associated impls M);
+
+// --- fail_todo_check_rewrite_constraints.carbon
 
 library "[[@TEST_NAME]]";
 
 import library "state_constraints";
 
 // TODO: Should fail since `2` can't be converted to the type of `I.Member`,
 // but constraints are not yet type checked.
```

Comment:
Has this TODO been resolved? If so, should we take `todo` out of the filename?

Reply:
Well, it now tries to do the type checking, but the bug in the member access machinery means it doesn't get the right result, so there is definitely still some "todo" left. I've removed `todo` from the filename, and updated the comment to clarify what still remains.

Reply:
I don't see those changes.

Reply:
Sorry, I fogot to push them. They are there now.